src/lib.rs

```diff
@@ -1,350 +1,440 @@
+where
+    T: Eq,
+    T: Ord,
+{
     if data.len() == 0 {
         return 1.0;
-    } 
+    }
     fn p_squared(count: usize, len: f32) -> f32 {
         let p = count as f32 / len;
         p * p
     }
     let len = data.len() as f32;
-    let mut count = HashMap::new();
-    for &value in data {
+    let mut count = BTreeMap::new();
+    for value in data.iter() {
         *count.entry(value).or_insert(0) += 1;
     }
     let counts: Vec<usize> = count.into_iter().map(|(_, c)| c).collect();
-    let indiv : Vec<f32> = counts.iter().map(|x| p_squared(*x, len)).collect();
-    let sum : f32 = indiv.iter().sum();
-    1.0 - sum
+    let indiv: Vec<f32> = counts.iter().map(|x| p_squared(*x, len)).collect();
+    let sum: f32 = indiv.iter().sum();
+    return 1.0 - sum;
 }
 
 /// The categorical accuracy of a dataset
-/// 
+///
 /// Returns a float where 1.0 is a perfectly accurate dataset
 /// ```
 /// use parsnip::categorical_accuracy;
-/// let pred = vec![0, 0, 0 , 1, 2];
-/// let actual = vec![1, 1, 1, 1, 2];
+/// let pred : Vec<u16> = vec![0, 0, 0 , 1, 2];
+/// let actual : Vec<u16> = vec![1, 1, 1, 1, 2];
 /// assert_eq!(categorical_accuracy(&pred, &actual), 0.4);
 /// ```
-pub fn categorical_accuracy(pred: &[u64], actual: &[u64]) -> f32 {
+pub fn categorical_accuracy<T>(pred: &[T], actual: &[T]) -> f32
+where
+    T: Eq,
+{
     assert_eq!(pred.len(), actual.len());
-    let bools =  pred.iter().zip(actual).map(|(x,y)| x == y);
-    let truthy : Vec<bool> =  bools.filter(|b| *b).collect();
-    truthy.len() as f32 / pred.len() as f32
+    let truthy = pred.iter().zip(actual).filter(|(x, y)| x == y).count();
+    return truthy as f32 / pred.len() as f32;
 }
 
-fn class_precision(pred: &[u64], actual: &[u64], class: u64) -> f32 {
+fn class_precision<T>(pred: &[T], actual: &[T], class: &T) -> f32
+where
+    T: Eq,
+{
     assert_eq!(pred.len(), actual.len());
-    let true_positives_map = pred.iter().zip(actual).map(|(p, a)| p == a && *p == class);
-    let true_positives = true_positives_map.filter(|b| *b).count() as f32;
-    let all_positives = pred.iter().map(|p| *p == class).filter(|b| *b).count() as f32;
+    //First, get the map of all true positives
+    let true_positives = pred
+        .iter()
+        .zip(actual)
+        .filter(|(p, a)| p == a && **p == *class)
+        .count() as f32;
+    let all_positives = pred.iter().filter(|p| **p == *class).count() as f32;
     if all_positives == 0.0 {
         0.0
     } else {
         true_positives / all_positives
     }
 }
 
-fn weighted_precision(pred: &[u64], actual: &[u64]) -> f32 {
+fn weighted_precision<T>(pred: &[T], actual: &[T]) -> f32
+where
+    T: Eq,
+    T: Ord,
+{
     assert_eq!(pred.len(), actual.len());
-    let mut classes : Vec<u64> = pred.into_iter().map(|x| *x).collect();
-    let mut class_weights = HashMap::new();
+    let mut classes: Vec<&T> = pred.into_iter().collect();
+    let mut class_weights = BTreeMap::new();
     classes.sort();
```

Comment:
If you can get rid of this, we can replace `Ord` with `Hash`, allowing us to use `HashMap` internally again.

Reply:
I brought them back to `Hash`, then used `HashSet` to deduplicate—sorting wasn't required. This should be a performance boost!